go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,157 @@
+	// +kubebuilder:validation:XValidation:rule="!has(self.ipamLifecycle) || has(self.subnets) && size(self.subnets) > 0", message="IPAMLifecycle is only supported when subnets are set"
+	// +required
+	Spec UserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetworkSpec.
+type UserDefinedNetworkSpec struct {
+	// topology describes network configuration.
+	// Allowed values are "Layer3", "Layer2", "LocalNet".
+	// Layer3 topology creates one logical switch per node - each with a different subnet - and a router interconnecting all the logical switches.
+	// Layer2 topology creates one logical switch shared by all nodes.
+	// LocalNet topology creates a cluster-wide logical switch connected to a physical network.
+	// +kubebuilder:validation:Required
+	// +required
+	Topology NetworkTopology `json:"topology"`
+
+	// role describes the network role in the pod.
+	// Allowed values are "Primary" and "Secondary".
+	// Primary network is automatically assigned to every pod created in the same namespace.
+	// Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+	// Only "Layer3" and "Layer2" topologies support "Primary" role.
+	// +kubebuilder:validation:Required
+	// +required
+	Role NetworkRole `json:"role"`
+
+	// mtu is the maximum transmission unit for a network.
+	// MTU is optional, if not provided, the globally configured value in OVN-Kubernetes (defaults to 1400) is used for the network.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65536
+	// from https://github.com/openshift/api/blob/0d46442e8df17a87cfcf6666ab19b28e88620b59/network/v1/types.go#L66
+	// +optional
+	MTU int32 `json:"mtu,omitempty"`
+
+	// subnets are used for the pod network across the cluster.
+	// Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed.
+	//
+	// When topology is "Layer3", the given subnet is split into smaller subnets for every node.
+	// To specify how the subnet should be split, the following format is supported for "Layer3" network:
+	// "10.128.0.0/16/24", which means that every host will get a "/24" subnet.
+	// If host subnet mask is not set (for example, "10.128.0.0/16"), it will be assigned automatically.
+	//
+	// For "Layer2" and "LocalNet" topology types, the format should match standard CIDR notation, without
+	// providing any host subnet mask.
+	// This field may be omitted for Secondary Layer2 and LocalNet networks.
+	// In that case the logical switch implementing the network only provides layer 2 communication,
+	// and users must configure IP addresses for the pods.
+	// Port security only prevents MAC spoofing
```

Comment:
Maybe write this so readers know this is a consequence of the subnets attribute being omitted. 

Something like:
```
	// This field may be omitted for Secondary Layer2 and LocalNet networks.
	// In that case, the logical switch implementing the network only provides layer 2 communication,
	// and users must configure IP addresses for the pods. As a direct consequence, Port security
        // only prevents MAC spoofing
```